src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -340,6 +340,125 @@ impl ScalarFunc {
         };
         Ok(Some(Cow::Owned(r.into_bytes())))
     }
+
+    #[inline]
+    pub fn substring_2_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        if pos == 0 {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+
+        let start = if mysql::has_unsigned_flag(self.children[1].get_tp().get_flag()) {
+            find_start(pos as u64 as usize, s.char_indices())
+        } else if pos > 0 {
+            find_start(pos as usize, s.char_indices())
+        } else {
+            let pos = 0_i64
+                .checked_sub(pos)
+                .map(|i| i as usize)
+                .unwrap_or_else(|| i64::max_value() as usize + 1);
+            find_start(pos, s.char_indices().rev())
+        };
+
+        if let Some(start) = start {
+            Ok(Some(Cow::Owned(s[start..].as_bytes().to_vec())))
+        } else {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+    }
+
+    #[inline]
+    pub fn substring_3_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        let len = try_opt!(self.children[2].eval_int(ctx, row));
+        let mut positive_search = true;
+
+        let pos = if mysql::has_unsigned_flag(self.children[1].get_tp().get_flag()) {
+            pos as u64 as usize
+        } else if pos > 0 {
+            pos as usize
+        } else {
+            positive_search = false;
+            0_i64
+                .checked_sub(pos)
+                .map(|i| i as usize)
+                .unwrap_or_else(|| i64::max_value() as usize + 1)
+        };
+
+        let len = if mysql::has_unsigned_flag(self.children[2].get_tp().get_flag()) {
+            len as u64 as usize
+        } else if len > 0 {
+            len as usize
+        } else {
+            0
+        };
+
+        if pos == 0 || len == 0 {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        let mut start = None;
+        let mut end = s.len();
+        let mut cnt = 1;
+        if positive_search {
+            for (i, _) in s.char_indices() {
+                if cnt > len && (cnt - len) >= pos {
+                    end = i;
+                    break;
+                }
+                if cnt == pos {
+                    start = Some(i);
+                }
+                cnt += 1;
+            }
+        } else {
+            let mut positions = VecDeque::with_capacity(len);
+            positions.push_back(end);
+            for (i, _) in s.char_indices().rev() {
```

Comment:
Same as below

Reply:
I updated the code using find here, but it looks a little weird, since it not only find the start or end position,  it also has side-effect  for other variables, what do you think?

Reply:
It's in any sense better.